File: mobile/lib/core/services/device_service.dart
import '../constants/app_constants.dart';
import 'api_client.dart';

class DeviceService {
  final ApiClient _apiClient = ApiClient();

  // Get device status
  Future<Map<String, dynamic>> getDeviceStatus() async {
    try {
      final response = await _apiClient.get(AppConstants.deviceStatusEndpoint);

      if (response.statusCode == 200 && response.data['success'] == true) {
        // Convert object to array of devices
        final Map<String, dynamic> deviceData = response.data['data'];
        final List<Map<String, dynamic>> devices = deviceData.entries.map((entry) {
          return {
            'deviceName': entry.key,
            'status': entry.value,
          };
        }).toList();
        
        return {'success': true, 'devices': devices};
      }

      return {
        'success': false,
        'message':
            response.data['message'] ?? 'Không thể tải trạng thái thiết bị',
      };
    } catch (e) {
      return {'success': false, 'message': 'Lỗi: ${e.toString()}'};
    }
  }

  // Control device
  Future<Map<String, dynamic>> controlDevice(
    String deviceName,
    String action,
  ) async {
    try {
      final response = await _apiClient.post(
        AppConstants.deviceControlEndpoint,
        data: {'deviceName': deviceName, 'action': action},
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        return {
          'success': true,
          'message': response.data['message'],
          'device': response.data['data'],
        };
      }

      return {
        'success': false,
        'message': response.data['message'] ?? 'Điều khiển thất bại',
      };
    } catch (e) {
      return {'success': false, 'message': 'Lỗi: ${e.toString()}'};
    }
  }

  // Get device control history
  Future<Map<String, dynamic>> getControlHistory({
    int page = 1,
    int limit = 20,
    String? deviceName,
  }) async {
    try {
      final queryParams = {'page': page.toString(), 'limit': limit.toString()};

      if (deviceName != null && deviceName.isNotEmpty) {
        queryParams['deviceName'] = deviceName;
      }

      final response = await _apiClient.get(
        AppConstants.deviceHistoryEndpoint,
        queryParameters: queryParams,
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        return {
          'success': true,
          'history': response.data['data']['history'],
          'pagination': response.data['data']['pagination'],
        };
      }

      return {
        'success': false,
        'message': response.data['message'] ?? 'Không thể tải lịch sử',
      };
    } catch (e) {
      return {'success': false, 'message': 'Lỗi: ${e.toString()}'};
    }
  }
}

File: Front End Web/src/services/device.service.ts
import apiClient from "./api.client";
import { API_ENDPOINTS } from "../config/api.config";

export interface DeviceControlRequest {
  deviceName:
    | "pump"
    | "fan"
    | "light"
    | "servo_door"
    | "servo_feed"
    | "led_farm"
    | "led_animal"
    | "led_hallway";
  action: "ON" | "OFF" | "AUTO";
  value?: number; // 0-255 for PWM, or 0-90/50 for servos
}

export interface DeviceStatusResponse {
  [key: string]: "ON" | "OFF" | "AUTO";
}

export interface DeviceHistoryParams {
  deviceName?: string;
  limit?: number;
  page?: number;
}

export const deviceService = {
  async controlDevice(
    data: DeviceControlRequest
  ): Promise<{ success: boolean; message: string }> {
    const response = await apiClient.post<{
      success: boolean;
      message: string;
    }>(API_ENDPOINTS.DEVICES.CONTROL, data);
    return response.data;
  },

  async getStatus(): Promise<DeviceStatusResponse> {
    const response = await apiClient.get<{
      success: boolean;
      data: DeviceStatusResponse;
    }>(API_ENDPOINTS.DEVICES.STATUS);
    return response.data.data;
  },

  async getHistory(params?: DeviceHistoryParams): Promise<any> {
    const response = await apiClient.get(API_ENDPOINTS.DEVICES.HISTORY, {
      params,
    });
    return response.data;
  },
};
